scripts/variantstore/wdl/GvsExtractCallsetPgen.wdl

```diff
@@ -0,0 +1,603 @@
+    String fq_ranges_cohort_vet_extract_table_name = "~{full_extract_prefix}__VET_DATA"
+    String fq_ranges_cohort_vet_extract_table = "~{fq_cohort_dataset}.~{fq_ranges_cohort_vet_extract_table_name}"
+
+    String fq_samples_to_extract_table = "~{fq_cohort_dataset}.~{full_extract_prefix}__SAMPLES"
+    Array[String] tables_patterns_for_datetime_check = ["~{full_extract_prefix}__%"]
+
+    Boolean emit_pls = false
+    Boolean emit_ads = true
+
+    String intervals_file_extension = if (zero_pad_output_pgen_filenames) then '-~{output_file_base_name}.interval_list' else '-scattered.interval_list'
+
+    if (!defined(git_hash) || !defined(gatk_docker) || !defined(cloud_sdk_docker) || !defined(variants_docker)) {
+        call Utils.GetToolVersions {
+            input:
+                git_branch_or_tag = git_branch_or_tag,
+        }
+    }
+
+    String effective_gatk_docker = select_first([gatk_docker, GetToolVersions.gatk_docker])
+    String effective_cloud_sdk_docker = select_first([cloud_sdk_docker, GetToolVersions.cloud_sdk_docker])
+    String effective_variants_docker = select_first([variants_docker, GetToolVersions.variants_docker])
+    String effective_git_hash = select_first([git_hash, GetToolVersions.git_hash])
+    call Utils.ScaleXYBedValues {
+        input:
+            interval_weights_bed = interval_weights_bed,
+            x_bed_weight_scaling = x_bed_weight_scaling,
+            y_bed_weight_scaling = y_bed_weight_scaling,
+            variants_docker = effective_variants_docker,
+    }
+
+    call Utils.GetBQTableLastModifiedDatetime as SamplesTableDatetimeCheck {
+        input:
+            project_id = project_id,
+            fq_table = fq_sample_table,
+            cloud_sdk_docker = effective_cloud_sdk_docker,
+    }
+
+    call Utils.GetNumSamplesLoaded {
+        input:
+            fq_sample_table = fq_sample_table,
+            project_id = project_id,
+            control_samples = control_samples,
+            sample_table_timestamp = SamplesTableDatetimeCheck.last_modified_timestamp,
+            cloud_sdk_docker = effective_cloud_sdk_docker,
+    }
+
+    Int effective_scatter_count = if defined(scatter_count) then select_first([scatter_count])
+                                  else if GetNumSamplesLoaded.num_samples < 100 then 100 # Quickstart
+                                       else if GetNumSamplesLoaded.num_samples < 1000 then 500
+                                            else if GetNumSamplesLoaded.num_samples < 5000 then 1000
+                                                 else if GetNumSamplesLoaded.num_samples < 20000 then 2000 # Stroke Anderson
+                                                      else if GetNumSamplesLoaded.num_samples < 50000 then 10000
+                                                           else if GetNumSamplesLoaded.num_samples < 100000 then 20000 # Charlie
+                                                                 else 34000
```

Comment:
so we would be in the 34000 case here... do we think that will actually work Cromwell-wise?

Reply:
That's a good question.  It's below the Terra maximum so it works there, but I haven't tried directly with a Cromwell instance.  I assume it's dependent on how the Cromwell instance is configured.

This logic would really benefit from a rework to be more robust.  From the tests I did, it seems like the number of shards needed is actually dependent on a combination of the number of samples you're extracting, the number of the samples in the full GVS callset, and the size of the interval list, and each of those has a pretty significant effect.  For example, I never successfully ran a whole genome extract for 100k or more samples (even with 34k shards), but I got up to 250k samples with the acaf_threshold interval list, and that was only using 20k shards (and it probably would've worked with even less).  I'm just not sure how best to apply that knowledge to improving this.

Reply:
I'd be happy to share any data I can to help improve it